src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs

```diff
@@ -18673,11 +18673,33 @@ public static Vector512<T> Invoke(Vector512<T> x)
         /// <summary>T.PopCount(x)</summary>
         internal readonly struct PopCountOperator<T> : IUnaryOperator<T, T> where T : IBinaryInteger<T>
         {
-            public static bool Vectorizable => false; // TODO: Vectorize
+            public static bool Vectorizable => sizeof(T) == 1 && (AdvSimd.IsSupported || PackedSimd.IsSupported);
+
             public static T Invoke(T x) => T.PopCount(x);
-            public static Vector128<T> Invoke(Vector128<T> x) => throw new NotSupportedException();
-            public static Vector256<T> Invoke(Vector256<T> x) => throw new NotSupportedException();
-            public static Vector512<T> Invoke(Vector512<T> x) => throw new NotSupportedException();
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static Vector128<T> Invoke(Vector128<T> x)
+            {
+                Debug.Assert(sizeof(T) == 1);
+
+                if (AdvSimd.IsSupported)
+                {
+                    return AdvSimd.PopCount(x.AsByte()).As<byte, T>();
+                }
+
+                if (PackedSimd.IsSupported)
+                {
+                    return PackedSimd.PopCount(x.AsByte()).As<byte, T>();
+                }
+
+                throw new NotSupportedException();
```

Comment:
Did you check the cost of doing it as something like this:
```csharp
Vector128<byte> c1 = Vector128.Create((byte)0x55);
Vector128<byte> c2 = Vector128.Create((byte)0x33);
Vector128<byte> c3 = Vector128.Create((byte)0x0F);

// We don't have a per element shuffle for byte on some platforms
//
// However, we do currently always have a 16-bit shift available and
// due to how the algorithm works, we don't need to worry about
// any bits that shift into the lower 8-bits from the upper 8-bits.

Vector128<byte> temp = x - ((x.AsUInt16() >> 1).AsByte() & c1);
temp = (temp & c2) + ((temp.AsUInt16() >> 2).AsByte() & c2);
return ((temp + (temp.AsUInt16() >> 4).AsByte()) & c3);
```

This is just a simplification of the software fallback algorithm used for `BitOperations.PopCount` for `int32` (since we only care about 8, not the full 32-bits)

On current main, this will generate something like the following for x86/x64:
```asm
vmovups  xmm0, xmmword ptr [reloc @RWD00]
vmovups  xmm1, xmmword ptr [rdx]
vpsrlw   xmm2, xmm1, 1
vpand    xmm2, xmm2, xmmword ptr [reloc @RWD16]
vpsubb   xmm1, xmm1, xmm2
vpand    xmm2, xmm1, xmm0
vpsrlw   xmm1, xmm1, 2
vpand    xmm0, xmm1, xmm0
vpaddb   xmm1, xmm0, xmm2
vpsrlw   xmm0, xmm1, 4
vpaddb   xmm0, xmm0, xmm1
vpand    xmm0, xmm0, xmmword ptr [reloc @RWD32]
vmovups  xmmword ptr [rcx], xmm0
mov      rax, rcx
```

Reply:
The same general setup should be possible to emulate for 16-bits, 32-bits, and 64-bits as well (just with the correctly adjusted algorithm).

Reply:
> Did you check the cost of doing it as something like this:

On my machine, with 256-bit vectors, that's still a few percent slower than just doing the scalar.

Reply:
Extending it out to Int32, it does measurably better but still not great: 2x the scalar with 256-bit vectors.

Reply:
And extending it out to Int64, it does horribly, at least on my machine: ~60x slower than scalar with 256-bit vectors.

Reply:
> On my machine, with 256-bit vectors, that's still a few percent slower than just doing the scalar.

I'm actually really surprised at that. I'd expect the `loop + extract + popcnt + insert` to be quite a bit more expensive, especially repeated 16 or 32 times.

`popcnt` alone is `3 cycles` with only a single available port, `pextrb` is `<=4 cycles` and takes 2 ports. `pinsrb` is `<= 4` cycles and also takes 2 ports. So the base cost for V128 should be `80-176 cycles` with almost no parallel dispatch opportunities. -- Notably `popcnt` is cheaper on AMD, but `pextrb` and `pinsrb` are more expensive. Costs haven't changed for newer Intel as compared to Skylake

Where-as the SIMD sequence should be `<=2 cycles` for `psrlw` (takes 2 ports, it is always 1 cycle taking 1 port for a constant shift amount), `1` cycle for `pand`, `psubb`, or `paddb` (1 port, 3 available), and `4-7 cycles` for `movups` or an embedded load. So the base cost should be: `10-13 cycles` (nearly 8x faster) if you factor in hoisting the constants or `22-34 cycles` if you only do a standalone execution (still over 2x faster). -- Costs remain the same on AMD and on newer Intel.

-------------------------

On my machine, I see the following for just a standalone call of `Popcnt` over a V128.
```csharp
BenchmarkDotNet v0.13.12, Windows 11 (10.0.22631.3085/23H2/2023Update/SunValley3)
AMD Ryzen 9 7950X, 1 CPU, 32 logical and 16 physical cores
.NET SDK 8.0.100
  [Host]     : .NET 8.0.0 (8.0.23.53103), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI
  DefaultJob : .NET 8.0.0 (8.0.23.53103), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI


| Method  | x                    | Mean       | Error     | StdDev    |
|-------- |--------------------- |-----------:|----------:|----------:|
| VPopcnt | <0, 1(...), 15> [54] |  0.2624 ns | 0.0241 ns | 0.0368 ns |
| Popcnt  | <0, 1(...), 15> [54] | 60.4811 ns | 0.0671 ns | 0.0524 ns |
```

I see almost identical numbers on my Intel Tiger Lake processor.

Reply:
I'm not able to repro what I was seeing before for bytes, so we can rack that up to user error.

I am, however, able to see the same regressions for long, and minimal speedups for int.

This is on Coffee Lake.

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.Numerics.Tensors;

BenchmarkSwitcher.FromAssembly(typeof(Tests).Assembly).Run(args);

[MemoryDiagnoser(false)]
[HideColumns("Job", "Error", "StdDev", "Median", "RatioSD")]
public class Tests
{
    private byte[] _bytes = Enumerable.Range(0, 10_000).Select(i => (byte)i).ToArray();
    private short[] _shorts = Enumerable.Range(0, 10_000).Select(i => (short)i).ToArray();
    private int[] _ints = Enumerable.Range(0, 10_000).ToArray();
    private long[] _longs = Enumerable.Range(0, 10_000).Select(i => (long)i).ToArray();

    private byte[] _bytesDest = new byte[10_000];
    private short[] _shortsDest = new short[10_000];
    private int[] _intsDest = new int[10_000];
    private long[] _longsDest = new long[10_000];

    [Benchmark]
    public void PopCountBytes() => TensorPrimitives.PopCount<byte>(_bytes, _bytesDest);

    [Benchmark]
    public void PopCountShorts() => TensorPrimitives.PopCount<short>(_shorts, _shortsDest);

    [Benchmark]
    public void PopCountInts() => TensorPrimitives.PopCount<int>(_ints, _intsDest);

    [Benchmark]
    public void PopCountLongs() => TensorPrimitives.PopCount<long>(_longs, _longsDest);
}
```

Before vectorizing PopCount:

| Method         | Mean     |
|--------------- |---------:|
| PopCountBytes  | 4.658 us |
| PopCountShorts | 2.952 us |
| PopCountInts   | 2.958 us |
| PopCountLongs  | 3.265 us |

After vectorizing PopCount using this approach:

| Method         | Mean         |
|--------------- |-------------:|
| PopCountBytes  |     373.9 ns |
| PopCountShorts |   1,108.4 ns |
| PopCountInts   |   2,403.9 ns |
| PopCountLongs  | 185,684.9 ns |

Reply:
For `long` that tracks since we don't have a `64-bit` shift and it has to be emulated instead, making it quite a bit more complex. I'd expect that `long` would be better handled via `32-bit popcnt` + `pairwise add` + `clear upper 32-bits`

For `int` that also tracks, since we'd be only be doing `4x extracts/inserts` and the improvement would be much more minimal on modern hardware. I expect that it would be greater on older hardware, due to the higher costs of extract/insert on those platforms.

Reply:
> I'd expect that long would be better handled via 32-bit popcnt + pairwise add + clear upper 32-bits

What instruction is "pairwise add" on xarch?

Reply:
Yep, `HorizontalAdd` on xarch and `AddPairwise` on Arm64.

You'd might want to measure on if it's better to do:
```csharp
Vector128<uint> tmp = Popcnt(x.AsUInt32());
tmp = Ssse3.HorizontalAdd(tmp, tmp);
return Sse2.UnpackLow(tmp, Vector128<uint>.Zero).AsUInt64();
```

or if it's better to do:
```csharp
Vector128<uint> tmp = Popcnt(x.AsUInt32());
tmp = Sse.UnpackLow(tmp, Vector128<uint>.Zero) + Sse.UnpackHigh(tmp, Vector128<uint>.Zero);
return tmp.AsUInt64();
```

`phadd` can have some hidden expenses on some platforms, while `punpckh` and `punpckl` typically do not. Both achieve the same general thing here (add upper/lower bits together, ensure upper bits are clear.

Reply:
> or if it's better to do:
> Sse.UnpackLow(tmp, Vector128<uint>.Zero) + Sse.UnpackHigh(tmp, Vector128<uint>.Zero)

Will that do the right thing?  e.g. if tmp was [1,2,3,4], won't that result in two vectors [1,0,2,0] and [3, 0,4,0], which when added together would result in [4,0,6,0], rather than the desired [3,0,7,0]?

Reply:
Ah right, it'd need a `Shuffle` instead. Probably not worth the complexity then.

Reply:
What's the equivalent of HorizontalAdd for Vector512? Avx512 doesn't appear to offer that.

Reply:
That said, I'm not sure any of this is particularly worthwhile right now. Both the horizontal add and an approach using SumAbsoluteDifferences on a byte-based popcount were effectively on par with the scalar on my machine, with both 128-bit and 256-bit vectors.

Reply:
> What's the equivalent of HorizontalAdd for Vector512? Avx512 doesn't appear to offer that.

Explicit shuffling.

> That said, I'm not sure any of this is particularly worthwhile right now.

That's fair for `int64`. For 2 operations in V128, I can completely believe it's effectively identical and only marginally faster for the 4 operations in V256`.

I'd expect it is more meaningful for 32-bit, but that's also not a priority to improve.